System/Random.hs

```diff
@@ -949,91 +952,90 @@ runSTGen :: RandomGen g => g -> (forall s . STGen g s -> ST s a) -> (a, g)
--- pseudo-random number generator. Returns only the resulting random value.
+-- pseudo-random number generator. Returns only the resulting pseudo-random
+-- value.
 --
 -- @since 1.2
 runSTGen_ :: RandomGen g => g -> (forall s . STGen g s -> ST s a) -> a
 runSTGen_ g action = fst $ runSTGen g action
 
--- | The default pseudo-random number generator.
-type StdGen = SM.SMGen
+-- | The standard pseudo-random number generator. Uses the SplitMix
+-- implementation provided by the
+-- [splitmix](https://hackage.haskell.org/package/splitmix) package.
+newtype StdGen = StdGen { unStdGen :: SM.SMGen }
+  deriving Show
 
 instance RandomGen StdGen where
-  next = SM.nextInt
-  genWord32 = SM.nextWord32
-  genWord64 = SM.nextWord64
-  split = SM.splitSMGen
-
+  next = second StdGen . SM.nextInt . unStdGen
+  genWord32 = second StdGen . SM.nextWord32 . unStdGen
+  genWord64 = second StdGen . SM.nextWord64 . unStdGen
+  split g = let (g1, g2) = SM.splitSMGen $ unStdGen g in (StdGen g1, StdGen g2)
 
-{- |
-The function 'mkStdGen' provides an alternative way of producing an initial
-generator, by mapping an 'Int' into a generator. Again, distinct arguments
-should be likely to produce distinct generators.
--}
+-- | Constructs a 'StdGen' deterministically.
 mkStdGen :: Int -> StdGen
-mkStdGen s = SM.mkSMGen $ fromIntegral s
+mkStdGen = StdGen . SM.mkSMGen . fromIntegral
 
 
 -- $uniform
+-- This library provides two type classes to generate pseudo-random values:
 --
--- @random@ has two type classes for generation of random numbers:
--- 'Uniform' and 'UniformRange'. One for generating every possible
--- value and another for generating every value in range. In other
--- libraries this functionality frequently bundled into single type
--- class but here we have two type classes because there're types
--- which could have instance for one type class but not the other.
---
--- For example: 'Integer', 'Float', 'Double' have instance for
--- @UniformRange@ but there's no way to define @Uniform@.
+-- *   'UniformRange' is used to generate a value of a datatype uniformly
+--     within an inclusive range.
+-- *   'Uniform' is used to generate a value of a datatype uniformly over all
+--     possible values of that datatype.
 --
--- Conversely there're types where @Uniform@ instance is possible
--- while @UniformRange@ is not. One example is tuples: @(a,b)@. While
--- @Uniform@ instance is straightforward it's not clear how to define
--- @UniformRange@. We could try to generate values that @a <= x <= b@
--- But to do that we need to know number of elements of tuple's second
--- type parameter @b@ which we don't have.
+-- Types may have instances for both or just one of 'UniformRange' and
+-- 'Uniform'. A few examples illustrate this:
 --
--- Or type could have no order at all. Take for example
--- angle. Defining @Uniform@ instance is again straghtforward: just
--- generate value in @[0,2π)@ range. But for any two pair of angles
```

Comment:
I quite liked the angle example, but I won't cry if it gets removed.

Reply:
I like this example also

Reply:
I've added it back.